### PR TITLE
Wayland backend

### LINKED ISSUE #1243: Our own backends for X11/Wayland
There are way too many limitations and tradeoffs that GTK/GDK have to support both of them, while not allowing us to do things that we want to do. Notable examples are separate-thread rendering (we have to use our own X11 connection to do that) and issues with absolute window positioning on Wayland (Wayland has `xdg_surface_set_window_geometry` which isn't used by GTK for some reason which leads to our popups being broken).

There are some major caveats on that way:
1) System dialogs (open/save file, folder selection, etc). We need to either still load GTK or have them out-of-proc somehow. The out-of-proc way will cause issues with making dialogs modal to the parent window on X11
2) Input methods. Right now we are relying on GTK do support them
3) Client-side window decorations. Some of the wayland shells are requiring applications to draw window frame, title and buttons, we need to support that somehow

### LINKED ISSUE #2338: Chrome OS?
Being an early-days WPF fan, I'm really excited about this project!

Maybe a silly question, but I was wondering, is Chrome OS a supported target platform?

### PR DESCRIPTION
**Pull request by affederaffe**
## What does the pull request do?
Add a Wayland backend for Linux.
There hasn't been any activity on the [previous PR](https://github.com/AvaloniaUI/Avalonia/pull/8003), which is why I'm opening one on my own.
At the moment, you have to manually generate the Wayland bindings because I haven't figured out how to do so on build:
`cd ./src/Linux/NWayland/src/NWayland.Scanner && dotnet run`
This PR is still WIP, help is greatly appreciated.

What currently works:
- [x] Windowing:
    - [x] Window creation
    - [x] Window modes: Normal, Maximized, Fullscreen
    - [x] Resizing
    - [x] Popups
    - [x] File Dialogs (native + managed fallback)
    - [x] Tray icons
    - [x] Client Side Decorations
- [ ] Keyboard:
    - [x] Basic input
    - [x] Key repetition
    - [ ] ITextInputMethodImpl with [zwp_text_input_v3](https://wayland.app/protocols/text-input-unstable-v3#zwp_text_input_v3)
- [x] Mouse:
    - [x] Basic handling
    - [x] Scrolling
    - [x] Theme cursors
    - [x] Custom cursors
    - [x] Animated cursors
- [x] Touch:
    - [x] Basic handling
- [x] Data sharing:
    - [x] Clipboard:
        - [x] Reading
        - [x] Writing
    - [x] DND
- [x] Rendering:
    - [x] Backends:
        - [x] EGL
        - [x] FrameBuffer
    - [x] Modes:
        - [x] Deferred
        - [x] Immediate
        - [x] Composition

## What is the current behavior?
Currently, XWayland is used in order to run Avalonia apps on Wayland systems.

## What is the updated/expected behavior with this PR?
Use Wayland instead of X11 on compatible systems.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
Some projects where moved to the Linux solution.

## Obsoletions / Deprecations
None

## Fixed issues
Fixes #1243 
Fixes #2338 (?)

### CODE REVIEW

**Comment by kekekeks** [comment]
If Wayland bindings are stable, I'll publish a packaged version on nuget

**Comment by kekekeks** [comment]
@affederaffe I believe we already have rudimentary CSD that are used for Win32 backend. Resize-grips are missing but other than that they are more or less usable even with X11.

**Comment by kekekeks** [comment]
Ah, you were already trying to enable them, nvm

**Comment by affederaffe** [comment]
Neither resize nor move seem to be currently implemented and the X11 backend doesn't support CSD either, so I thought I'd disable them for now.

**Comment by affederaffe** [comment]
@kekekeks Any plans on releasing NWayland on NuGet?

**Review by jmacato — CHANGES_REQUESTED** [review]
@affederaffe @kekekeks this doesnt seem to work on GNOME coz i got this message

```
Unhandled exception. System.NotSupportedException: Unable to find zxdg_decoration_manager_v1 in the registry
   at Avalonia.Wayland.WlRegistryHandler.Bind[T](IBindFactory`1 factory, String interface, Int32 version) in /var/home/strix/RiderProjects/AvaloniaM/src/Linux/Avalonia.Wayland/WlRegistryHandler.cs:line 59
   at Avalonia.Wayland.WlRegistryHandler.BindRequiredInterface[T](IBindFactory`1 factory, String interface, Int32 version) in /var/home/strix/RiderProjects/AvaloniaM/src/Linux/Avalonia.Wayland/WlRegistryHandler.cs:line 49
   at Avalonia.Wayland.AvaloniaWaylandPlatform..ctor(WaylandPlatformOptions options) in /var/home/strix/RiderProjects/AvaloniaM/src/Linux/Avalonia.Wayland/AvaloniaWaylandPlatform.cs:line 37
   at Avalonia.AvaloniaWaylandPlatformExtensions.<>c__0`1.<UseWayland>b__0_0() in /var/home/strix/RiderProjects/AvaloniaM/src/Linux/Avalonia.Wayland/AvaloniaWaylandPlatform.cs:line 156
   at Avalonia.Controls.AppBuilderBase`1.Setup() in /var/home/strix/RiderProjects/AvaloniaM/src/Avalonia.Controls/AppBuilderBase.cs:line 280
   at Avalonia.Controls.AppBuilderBase`1.SetupWithLifetime(IApplicationLifetime lifetime) in /var/home/strix/RiderProjects/AvaloniaM/src/Avalonia.Controls/AppBuilderBase.cs:line 148
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /var/home/strix/RiderProjects/AvaloniaM/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 208
   at ControlCatalog.NetCore.Program.Main(String[] args) in /var/home/strix/RiderProjects/AvaloniaM/samples/ControlCatalog.NetCore/Program.cs:line 103
```

`cd ./src/Linux/NWayland/src/NWayland.CodeGen && dotnet run` should be updated to ` cd ./src/Linux/NWayland/src/NWayland.Scanner && dotnet run`

**Comment by kekekeks** [comment]
@jmacato you need to use a desktop environment that supports server decorations. Basically not GNOME.

**Comment by jmacato** [comment]
@affederaffe im still getting the same error in gnome on the latest commit:
image

**Comment by kekekeks** [comment]
Yes, your desktop environment needs to support Server Side Decorations. Please, check with desktop environment that supports `zxdg_decoration_manager_v1`

**Comment by affederaffe** [comment]
Decorations seem to work on Gnome now, though I noticed that popups and drag&drop are completely broken on it, even though they work perfectly fine on KDE.

**Comment by jmacato** [comment]
image
This is wild :D but it seems like that we need a way to add CSD on windows that didnt specify it. I know GNOME is a bit thickheaded but it's a major DE that we cannot ignore.

**Comment by jmacato** [comment]
@affederaffe this is so amazing :D i can do DnD on linux now :D thanks so much for this effort!

btw, some more bug reports if you dont mind:

image

Some windows still doesnt have CSD for some reason, like the DevTools window for example

**Comment by jmacato** [comment]
@affederaffe @kekekeks im curious, is it possible to make a RenderTimer that is synchronized to the Compositor's refresh rate?

**Comment by kekekeks** [comment]
You can ask the compositor to send you a callback when it thinks that's a good idea to start rendering a new frame. However for that to work each window would have its own render timer. While being possible to refactor our rendering to use multiple render timers, I'm afraid the UI thread animations like their global variables a bit too much, so we can't do that.

**Comment by jmacato** [comment]
@kekekeks by global variables you mean the inherited Clocks?

**Comment by kekekeks** [comment]
Yes, UI thread animations expect to have some global clock to function.

**Comment by jmacato** [comment]
@kekekeks that can be refactored yes, inherently we only need a way to hand over some kind of timer to the clocks

**Comment by kekekeks** [comment]
>each window/surface can have one timer each and it should work fine

Animations can tick on brushes in resource dictionaries that are shared by multiple windows.

**Comment by Conan-Kudo** [comment]
You might want to look into [libdecor](https://gitlab.gnome.org/jadahl/libdecor) for handling SSD-less compositors (basically just Mutter from GNOME).

**Comment by kekekeks** [comment]
Last time I've checked libdecoration required glib and wasn't really thread safe, which is important because Avalonia does event handling/layout and rendering on different threads.

**Comment by jmacato** [comment]
@kekekeks i like having our own CSD implementation better tbh. That way users can modify it so without much fuss.

**Review by jmacato — CHANGES_REQUESTED** [review]
image

I cant seem to build this in a fresh clone

**Comment by affederaffe** [comment]
@jmacato You need to add
`<ItemGroup Label="InternalsVisibleTo">
        <InternalsVisibleTo Include="NWayland.Protocols.Plasma, PublicKey=$(AvaloniaPublicKey)" />
        <InternalsVisibleTo Include="NWayland.Protocols.Wlr, PublicKey=$(AvaloniaPublicKey)" />
    </ItemGroup>`
for the time being, that'll be fixed once the bindings are published on NuGet.

**Comment by jmacato** [comment]
@affederaffe im getting this error now as well, fixed the public sigs though with your comment
image

**Comment by affederaffe** [comment]
@jmacato Update the NWayland submodule and regenerate the bindings

**Review by jmacato — CHANGES_REQUESTED** [review]
@affederaffe please fix the latest merge conflict with master, i tested this PR and it's working great so far :D 

Just noticed some issues:

- Copy/Paste/Clipboard doesnt seem to parse and decode Unicode characters.
- Frame rate is still locked to 60fps for high refresh rate monitors but im not sure if that's fixable in this current PR.
- OpenGL doesnt seem to work on dedicated GPU only setup.

I dont think we can merge until we can get a new NWayland package from @kekekeks. Or at least make NWayland use sourcegens instead, not sure.

Anyways this is a very much needed backend since X11 going to be endangered species in the next few years.

**Comment by affederaffe** [comment]
I opened a PR with my fixes for NWayland [here](https://github.com/kekekeks/NWayland/pull/1).

- Should Copy/Paste decode entered unicode codes? As far as I know hard-coded charachters are decoded correctly and only those entered by the user are escaped, just like with X11.
- The framerate can easily be unlocked once there's a solution for brushes ticking in ResourceDictionaries as discussed above.
- This seems to be an issue with shared contexts with the EGL backend and doesn't work on X11 either. I can try to fix this, though I have no knowledge on this so I'd appreciate a little help.

**Comment by jmacato** [comment]
@affederaffe im waiting for @kekekeks to review your PR on NWayland's repo. After he publishes the nuget we can get this PR merged :)

File: src/Linux/Avalonia.FreeDesktop/LibC.cs
Comment (jmacato) [inline src/Linux/Avalonia.FreeDesktop/LibC.cs:55]: I wonder if we can use `nuint` on this?

**Comment by jmacato** [comment]
@affederaffe Seems like the FPS overlay is not displaying correctly on Gnome?
image

**Comment by squidink7** [comment]
I just tried this on Arch Linux KDE, and found issues related to maximising. The window would only maximise to a small portion of the actual screen size, and would become even smaller when any other window was focused
image
I also needed to remove the InternalsVisibleTo lines in the NWayland project file or no apps would build.
Very nice work though, other than that everything was working well as far as I saw!

**Comment by robloo** [comment]
Hope there is discussion with the core team on this and a plan for merging. Really should have been merged in before the RC in my opinion (so by preview 5). It needs a chance for good testing before release.

**Comment by timunie** [comment]
Hey @affederaffe ,

I want to reach out for you as 11.0 is going to be closer and this PR is a really huge PR. We really appreciate your efforts and it is on our todo-list to get it merged. But most likely it will not be included in 11.0, as 11.0 has already a lot of new stuff inside. We plan to add this feature in a 11.x release after the major 11.0 bugs are ironed out. 

I hope for your understanding 💐 


Happy coding

Tim

**Comment by affederaffe** [comment]
I completely understand, also there are still some bugs left I haven't had the time to fix yet and probably won't in the near future, so it's definitely for the better. Still, I plan to work on this PR eventually, it is _not_ abandoned.

**Comment by timunie** [comment]
Okay, so we are on the same route. That's great. I think it is a good idea to merge master as you already do it in order to not get too much out of sync. Anything else can be done later 👍 

Have a nice day.

**Comment by kekekeks** [comment]
To be clear: we are planning to extract any changes to Avalonia core libs from this PR and merge them before 11.0. That way the rest of Wayland support could be merged later in 11.x.

**Comment by jgcodes2020** [comment]
If it's helpful, here's an alternative Wayland scanner: https://github.com/X9VoiD/WaylandSharp

**Comment by nicola36631** [comment]
@affederaffe is there some way the community can help with this PR? What is still missing/not working at the moment?

**Comment by kekekeks** [comment]
Note that the way CSD are implemented kinda interferes with Window's Width/Height controlling the client area size. So it probably can't be merged as is and would require some major changes in either TopLevel or composition code.

**Comment by jgcodes2020** [comment]
> Note that the way CSD are implemented kinda interferes with Window's Width/Height controlling the client area size. So it probably can't be merged as is, proper CSD would require some major changes in either TopLevel or composition code.

Proper CSD might be worth splitting into a separate PR, since there are already a lot of changes here. All that matters for a Wayland backend is that it runs on GNOME (which has no current or planned `xdg-decoration` support).

**Comment by kekekeks** [comment]
Gnome is pretty much the only "very special child" that needs us to implement CSD. Without CSD it won't show any window frame.

We could merge Wayland without CSD and ignore gnome for now since it would enable some embedded device scenarios that are running fullscreen anyways.

**Comment by nicola36631** [comment]
> > Note that the way CSD are implemented kinda interferes with Window's Width/Height controlling the client area size. So it probably can't be merged as is, proper CSD would require some major changes in either TopLevel or composition code.
> 
> Proper CSD might be worth splitting into a separate PR, since there are already a lot of changes here. All that matters for a Wayland backend is that it runs on GNOME (which has no current or planned `xdg-decoration` support).


I think that supporting GNOME and KDE Plasma (best looking) and XFCE (most lightweight) is quite important as they are "the top dogs" of desktop environments on Linux.


Perhaps improving Client Side Decorations is an issue that can be moved to the 11.x milestone?

**Comment by Doomsdayrs** [comment]
Closed??

**Comment by zoujiaqing** [comment]
Wow...

**Comment by robloo** [comment]
Hopefully this is just closed because the strategy was changed a bit for 11.0

1. Pull out the core changes it needs and get those in for 11.0
2. Finish and merge this in the 11.x series in a separate PR

If this was closed because it is no longer developed that is most unfortunate. It should have been merged as-is and provided a good base for everyone to test and patch from. 95% was more than OK in my opinion for a previously unsupported backend.

**Comment by affederaffe** [comment]
I'm just starting to rebase my patches onto master and I decided it's easier to do so on a new branch. That comes with the advantage of a clean master branch and a feature branch, like it should've been from the start. The wayland backend still needs some work and I don't have that much time in the coming weeks, after that I do though.

**Pull request closed without merging**
